### DISCUSSION TITLE
Okayish text-to-image generation results on COCO

### DISCUSSION THREAD

**Discussion by kobiso** [lucidrains/DALLE-pytorch#181] [Show and tell]
## DALLE on COCO
- I trained DALLE + VQGAN ([Taming transformers](https://github.com/CompVis/taming-transformers)) with COCO dataset and got okayish generation results.
- Note that the model is not trained on Deepspeed DDP, but trained on horovod DDP.
  - dalle-pytorch version [0.7.2](https://github.com/lucidrains/DALLE-pytorch/releases/tag/0.7.2) is used.
- Pretrained model is available: https://github.com/kobiso/DALLE-reproduction

#### Text to image generation and re-ranking by CLIP (from training text)
![image](https://user-images.githubusercontent.com/12422441/114018391-23f33080-98a8-11eb-8e98-493720ffc64c.png)

<details>
<summary>Click here to see more results</summary>

![image](https://user-images.githubusercontent.com/12422441/114018417-2ce40200-98a8-11eb-97d2-781ddc3e3724.png)
![image](https://user-images.githubusercontent.com/12422441/114018443-32414c80-98a8-11eb-867a-9c55e1b21868.png)

</details>

#### Generate rest of image based on the given cropped image (from training text)
![image](https://user-images.githubusercontent.com/12422441/114018463-38372d80-98a8-11eb-8c26-a1558d5483ec.png)

<details>
<summary>Click here to see more results</summary>

![image](https://user-images.githubusercontent.com/12422441/114018478-3c634b00-98a8-11eb-877d-08cdc4101e74.png)
![image](https://user-images.githubusercontent.com/12422441/114018495-408f6880-98a8-11eb-83ea-78b6902afded.png)

</details>

#### Text to image generation with custom text
![image](https://user-images.githubusercontent.com/12422441/114018982-d4613480-98a8-11eb-85cf-3965d05e91f6.png)

<details>
<summary>Click here to see more results</summary>

![image](https://user-images.githubusercontent.com/12422441/114019008-dcb96f80-98a8-11eb-8503-01a4b440e4fa.png)

</details>

#### VAE
- Pretrained VQGAN

#### DALLE
- dim = 256
- text_seq_len = 80
- depth = 32
- heads = 8
- dim_head = 64
- reversible = 0
- attn_types = (full,sparse)

#### Optimization
- Optimizer: Adam
- Learning rate: start with 0.00045 and perform `ReduceLROnPlateau`
  - Learning rate has to be scaled by the number of gpus: 0.00045 * gpus
- Apply loss weighting
- Gradient clipping is necessary, unless NaN will occur
- Batch size: 64 * 8 (gpus)
   - I used [horovod](https://github.com/horovod/horovod) for multi-gpus training
- The model is not converged, so it could be better when it converged.

### Opinion
- It seems generation from the training text is okay, but generation from custom text is poor.
- When CUB is only dealing with birds, COCO has more kinds of objects with complicated backgrounds.
  - So, training a generalized DALLE requires much more number of datasets.

**Comment by mhh0318** [comment]
Cool, how about your bpe code book? you train another one with COCO or just use the same as your CUB200 experiments?

**Reply by mhh0318** [reply]
Got the new one from your project! Awesome

**Comment by afiaka87** [comment]
In the image mask example - did you use an image from the training set as the mask? Or from elsewhere? @kobiso

**Reply by kobiso** [reply]
@afiaka87 I used images from the training sets :)

**Comment by JeniaJitsev** [comment]
Looks not bad at all, congratulations on that! Can you give us an idea how long training took until the observed state, with 8 GPUs that you are using?

**Reply by kobiso** [reply]
The model is not fully converged one (I accidently stopped training in the middle).
So, it took 3 days and 5 hours with 8 GPUs.
When we wait the model to converge, it should be around a week i guess.

**Comment by GuoxingY** [comment]
The results look great! Could you share the loss during the training?

**Reply by kobiso** [reply]
Sure :)

![image](https://user-images.githubusercontent.com/12422441/118479376-e8237480-b74b-11eb-81d4-554c8a0d2d4e.png)

**Reply by GuoxingY** [reply]
Thanks so much ! Did you use the VQGAN pretrained on Imagenet and  train transformer on COCO with split 2017 ? I tried to train on COCO with split 2014 with your config but did not get the results as good as yours.
